tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,240 @@
+  ) =
+    for node in nodes:
+      node.subscribe(topic, handler)
+    echo "Subscribed all nodes to the topic: ", topic
+
+  proc commonUnsubscribe(
+      nodes: seq[TestGossipSub],
+      topic: string,
+      handler: proc(topic: string, data: seq[byte]) {.async.},
+  ) =
+    for node in nodes:
+      node.unsubscribe(topic, handler)
+    echo "Unsubscribed all nodes from the topic: ", topic
+
+  # Simulate the `SUBSCRIBE` to the topic and check proper handling in the mesh and gossipsub structures
+  asyncTest "handle SUBSCRIBE to the topic":
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    subscribeToTopics(gossipSub, @[topic])
+
+    check gossipSub.topics.contains(topic)
+
+    check gossipSub.gossipsub[topic].len() == 5
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Simulate an UNSUBSCRIBE to the topic and check if the topic is removed from the relevant data structures but remains in gossipsub
+  asyncTest "handle UNSUBSCRIBE to the topic":
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    subscribeToTopics(gossipSub, @[topic])
+
+    unsubscribeFromTopics(gossipSub, @[topic])
+
+    check topic notin gossipSub.topics
+    check topic notin gossipSub.mesh
+    check topic in gossipSub.gossipsub
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test subscribing and unsubscribing multiple topics
+  asyncTest "handle SUBSCRIBE and UNSUBSCRIBE multiple topics":
+    let topics = ["topic1", "topic2", "topic3"].toSeq()
+    let (gossipSub, conns) = setupGossipSub(topics, 5)
+
+    subscribeToTopics(gossipSub, topics)
+
+    check gossipSub.topics.len == 3
+    for topic in topics:
+      check gossipSub.gossipsub[topic].len() == 5
+
+    unsubscribeFromTopics(gossipSub, topics)
+
+    for topic in topics:
+      check topic notin gossipSub.topics
+      check topic notin gossipSub.mesh
+      check topic in gossipSub.gossipsub
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test ensuring that the number of subscriptions does not exceed the limit set in the GossipSub parameters
+  asyncTest "subscription limit test":
+    let topicCount = 15
+    let gossipSubParams = 10
+    let topicNames = toSeq(mapIt(0 .. topicCount - 1, "topic" & $it))
+
+    let (gossipSub, conns) = setupGossipSub(topicNames, 0)
+
+    gossipSub.topicsHigh = gossipSubParams
+
+    for topic in topicNames:
+      if gossipSub.topics.len < gossipSub.topicsHigh:
+        gossipSub.subscribe(
+          topic,
+          proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+            discard
+          ,
+        )
+      else:
+        check gossipSub.topics.len == gossipSub.topicsHigh
+
+    check gossipSub.topics.len <= gossipSub.topicsHigh
+    check gossipSub.topics.len == gossipSub.topicsHigh
```

Comment:
You don't need both checks, just one is enough. Please, keep the more exhaustive one.